getting-started/pages/ssr-example.tsx

```diff
@@ -1,3 +1,122 @@
-export default function ssrExample() {
-  return <></>;
+import { GetServerSideProps } from "next";
+import { useState } from "react";
+import {
+  CausalClient,
+  ImpressionJSON,
+  createQuery,
+  SelectFeatures,
+  toImpression,
+} from "../causal";
+import { RatingWidget } from "../components/RatingWidget";
+
+/****
+ **** Please read react-example.tsx first, it provides context necessary for this example
+ ***/
+
+// In Next.js to pre-render a page server side you define the function getServerSideProps
+export const getServerSideProps: GetServerSideProps<SSRProps> = async (
+  context
+) => {
+  // Get the product to display.
+  const product = products[context.query.pid as keyof typeof products];
+  if (product == undefined) {
+    return {
+      redirect: { destination: "ssr-example?pid=iphone" },
+    } as any; /* eslint-disable-line */ // cast OK, render will not be called
+  }
+
+  // Create a query using createQuery instead of queryBuilder()
+  // More on this in a moment
+  const query = createQuery<RatingFeatures>({
+    RatingBox: { product: product.name },
+  });
+
+  // request the impression with requestImpression
+  // we are not in a react render, so not using the react hook (useImpression)
+  // this is async code, so await the result
+  const { impression, error } = await causalClient.requestImpression(query);
+
+  if (error) {
+    console.log(
+      "There is no impression server running yet, but it still works! " +
+        "Causal is resilient to network and backend outages because the defaults are compiled in 😃."
+    );
+  }
+
+  // Next.js requires that SSR props be serializable as JSON
+  // So we have to convert the impression to JSON, and convert it back when
+  // we render.
+  const props: SSRProps = { product, json: impression.toJSON() };
+  return { props };
+};
+
+// createQuery() vs queryBuilder()
+//
+// When using queryBuilder(), the type of the query and the feature arguments
+// are built at the same time
+//
+// Sometimes we need to define the type first, and then use it later to both
+// create a query and also to pass data around.
+//
+// For this use case, SelectFeatures defines the type and createQuery
+// (used in gerServerSideProps) create the query
+//
+// SelectFeatures takes in the feature names to use as a generic parameter
+// i.e:  "Feature1" | "Feature2" | "Feature3"
+//
+// In VSCode, Use a double quote (") to trigger automplete within this <>
+type RatingFeatures = SelectFeatures<"RatingBox">;
```

Comment:
It's a bit weird that this is below it's actual usage in the file.

Reply:
Yes. I did it for the narrative flow of the comments, but I'm open to suggestions.